File: typescript-nodejs/architecture-shell-cqrs/src/IUnitOfWork.ts
/**
 * IUnitOfWork - Transaction boundary abstraction for commands
 *
 * Provides transaction management for state-changing operations.
 * Integrates with database-specific transaction mechanisms (e.g., TypeORM, Sequelize).
 *
 * @remarks
 * Behavior:
 * - beginTransaction(): Opens new transaction if none active; throws if transaction provider unavailable (fail fast per BR-006)
 * - hasActiveTransaction: Returns true if transaction currently active (prevents nested transaction attempts)
 * - commit(): Persists changes; called by UnitOfWork behavior on successful handler completion (including Result.Failure business errors)
 * - rollback(): Discards changes; called by UnitOfWork behavior on exception
 *
 * Constraints:
 * - UnitOfWork behavior MUST check hasActiveTransaction before calling beginTransaction()
 * - Transaction context propagates via AsyncLocalStorage<Transaction>
 * - Query handlers MUST NOT trigger transaction opening
 *
 * @example
 * ```typescript
 * export class TypeOrmUnitOfWork implements IUnitOfWork {
 *   private queryRunner?: QueryRunner;
 *
 *   get transactionId(): string {
 *     return this.queryRunner?.connection.driver.transactionId ?? '';
 *   }
 *
 *   get hasActiveTransaction(): boolean {
 *     return this.queryRunner?.isTransactionActive ?? false;
 *   }
 *
 *   async beginTransaction(signal: AbortSignal): Promise<void> {
 *     if (this.hasActiveTransaction) return; // Reuse existing
 *     this.queryRunner = this.connection.createQueryRunner();
 *     await this.queryRunner.connect();
 *     await this.queryRunner.startTransaction();
 *   }
 *
 *   async commit(signal: AbortSignal): Promise<void> {
 *     if (!this.hasActiveTransaction) return;
 *     await this.queryRunner!.commitTransaction();
 *     await this.queryRunner!.release();
 *   }
 *
 *   async rollback(signal: AbortSignal): Promise<void> {
 *     if (!this.hasActiveTransaction) return;
 *     await this.queryRunner!.rollbackTransaction();
 *     await this.queryRunner!.release();
 *   }
 * }
 * ```
 *
 * @see {@link UnitOfWorkBehavior} - Pipeline behavior that manages transaction lifecycle
 */
export interface IUnitOfWork {
  /**
   * Unique identifier for the current transaction
   * Used for correlation in logs and telemetry per NFR-002
   */
  readonly transactionId: string;

  /**
   * Indicates whether a transaction is currently active
   * Used to prevent nested transaction attempts
   */
  readonly hasActiveTransaction: boolean;

  /**
   * Opens a new transaction if none is active
   *
   * @param signal - Cancellation signal
   * @throws If transaction provider is unavailable (e.g., connection pool exhausted, database down)
   */
  beginTransaction(signal: AbortSignal): Promise<void>;

  /**
   * Commits the active transaction, persisting all changes
   *
   * @param signal - Cancellation signal
   * @throws If commit fails (e.g., constraint violation, deadlock)
   */
  commit(signal: AbortSignal): Promise<void>;

  /**
   * Rolls back the active transaction, discarding all changes
   *
   * @param signal - Cancellation signal
   */
  rollback(signal: AbortSignal): Promise<void>;
}

File: csharp-dotnet/src/Architecture.Shell.Cqrs/IUnitOfWork.cs
namespace Architecture.Shell.Cqrs;

/// <summary>
/// Transaction boundary abstraction for commands.
/// Provides transaction lifecycle management with support for nested command detection.
/// </summary>
/// <remarks>
/// <para><strong>Behavioral Requirements:</strong></para>
/// <list type="bullet">
/// <item>Commands MUST execute within transaction boundaries (enforced by UnitOfWorkBehavior)</item>
/// <item>Queries MUST NOT open transactions (no-op in UnitOfWorkBehavior)</item>
/// <item>Nested commands MUST reuse active transaction (check <see cref="HasActiveTransaction"/>)</item>
/// <item>Transaction commit occurs on successful handler completion (including Result.Failure business errors)</item>
/// <item>Transaction rollback occurs on exception (infrastructure errors only)</item>
/// <item>BeginTransactionAsync MUST fail fast when transaction provider unavailable (per BR-006)</item>
/// </list>
///
/// <para><strong>Usage Pattern:</strong></para>
/// <code>
/// // In UnitOfWorkBehavior
/// if (request is ICommand &amp;&amp; !_unitOfWork.HasActiveTransaction)
/// {
///     await _unitOfWork.BeginTransactionAsync(ct);
///     try
///     {
///         var response = await next();
///         await _unitOfWork.CommitAsync(ct); // Even if response is Result.Failure
///         return response;
///     }
///     catch (Exception)
///     {
///         await _unitOfWork.RollbackAsync(ct);
///         throw;
///     }
/// }
/// </code>
/// </remarks>
public interface IUnitOfWork
{
    /// <summary>
    /// Gets the transaction ID for correlation with telemetry logs.
    /// Populated after <see cref="BeginTransactionAsync"/> is called.
    /// </summary>
    /// <remarks>
    /// REQUIRED for all command executions per NFR-002 and BR-002.
    /// Used by TelemetryBehavior to log transaction context.
    /// </remarks>
    Guid TransactionId { get; }

    /// <summary>
    /// Indicates whether a transaction is currently active.
    /// Used by UnitOfWorkBehavior to detect nested commands and prevent nested transactions.
    /// </summary>
    bool HasActiveTransaction { get; }

    /// <summary>
    /// Opens a new database transaction.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token for early termination.</param>
    /// <exception cref="InvalidOperationException">
    /// When transaction provider is unavailable (connection pool exhausted, database offline).
    /// Per BR-006, this MUST fail fast without retry.
    /// </exception>
    /// <remarks>
    /// MUST generate new <see cref="TransactionId"/> and set <see cref="HasActiveTransaction"/> to true.
    /// </remarks>
    Task BeginTransactionAsync(CancellationToken cancellationToken);

    /// <summary>
    /// Commits the active transaction, persisting all changes.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token for early termination.</param>
    /// <remarks>
    /// Called by UnitOfWorkBehavior when handler completes successfully.
    /// This includes scenarios where handler returns Result.Failure (business errors are valid state).
    /// </remarks>
    Task CommitAsync(CancellationToken cancellationToken);

    /// <summary>
    /// Rolls back the active transaction, discarding all changes.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token for early termination.</param>
    /// <remarks>
    /// Called by UnitOfWorkBehavior when handler or behavior throws exception (infrastructure errors only).
    /// MUST be logged BEFORE rollback execution per BR-002 to ensure TransactionId correlation if rollback itself fails.
    /// </remarks>
    Task RollbackAsync(CancellationToken cancellationToken);
}
